pkg/maps/datametrics/datametrics.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2016-2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datametrics
+
+import (
+	"fmt"
+	"unsafe"
+
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/logging"
+)
+
+var log = logging.DefaultLogger
+
+const (
+	// MapName for metrics map.
+	MapName = "cilium_metrics"
+	// MaxEntries is the maximum number of keys that can be present in the
+	// Metrics Map.
+	MaxEntries = 65536
+)
+
+// Key must be in sync with struct metrics_key in <bpf/lib/common.h>
+type Key struct {
+	reason   uint8
+	dir      uint8
+	reserved [3]uint16
+}
+
+// Value must be in sync with struct metrics_value in <bpf/lib/common.h>
+type Value struct {
+	count uint64
+	bytes uint64
+}
+
+// String converts the key into a human readable string format
+func (k Key) String() string {
+	return fmt.Sprintf("reason:%d dir:%d", k.reason, k.dir)
+}
+
+// GetKeyPtr returns the unsafe pointer to the BPF key
+func (k Key) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(&k) }
+
+// NewValue returns a new empty instance of the structure representing the BPF
+// map value
+func (k Key) NewValue() bpf.MapValue { return &Value{} }
```

Comment:
Please use pointer receivers here. We hit issues some time ago with non-pointer receivers in map keys.

(Looks like ipcache needs the same fix, filed https://github.com/cilium/cilium/issues/4192)

Reply:
will do

Reply:
I am not really using this functionality, so removing it altogether.

Reply:
I checked and this is the definition of `//NewValue creates a new bpf.MapValue.
func (k *CtKey6) NewValue() bpf.MapValue { return &CtEntry{} }` in all other places... 
I think this function needs to be declared with this same syntax to properly identify the `key` interface? Please search `NewValue` .

Reply:
Yes, all of the functions with the Key receiver will need to be pointers. ie

`func (k *Key) NewValue() bpf.MapValue { return &Value{} }`,
not
`func (k Key) NewValue() bpf.MapValue { return &Value{} }`